File: Data Structures & Algorithms/islands-and-treasure/submission-15.py
class Solution:
    def islandsAndTreasure(self, grid: List[List[int]]) -> None:
        directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        ROWS, COLS = len(grid), len(grid[0])
        INF = 2147483647
        
        def bfs(r, c):
            q = deque([(r, c)])
            visit = set([(r, c)])
            length = 1
            while q:
                for i in range(len(q)):
                    row, col = q.popleft()
                    for dr, dc in directions:
                        nr, nc = row + dr, col + dc
                        if nr < 0 or nr >= ROWS or nc < 0 or nc >= COLS or (nr, nc) in visit or grid[nr][nc] == -1:
                            continue
                        # if grid[nr][nc] == -1:
                        #     continue
                        if grid[nr][nc] == 0:
                            return length
                        visit.add((nr, nc))
                        q.append((nr, nc))
                length += 1
            return INF
        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == INF: 
                    grid[r][c] = bfs(r, c)

File: Data Structures & Algorithms/interleaving-string/submission-15.cs
public class Solution {
    public bool IsInterleave(string s1, string s2, string s3) {
        int r = s1.Length, c = s2.Length;
        if(r + c != s3.Length) {
            return false;
        }
        bool[,] dp = new bool[r+1,c+1];
        dp[r,c] = true;
    
        for(int i = r; i >= 0; i--) {
            for(int j = c; j >= 0; j--) {
                if (i < r && s1[i] == s3[i+j] && dp[i+1,j]) {
                    dp[i,j] = true;
                }
                if (j < c && s2[j] == s3[i+j] && dp[i,j+1]) {
                    dp[i,j] = true;
                }
            }
        }
        return dp[0,0];
    }
}
